Add stable_sort_by_key (#1692)
oneapi-src/oneDPL: documentation/library_guide/parallel_api/additional_algorithms.rst

    search sequence: [0, 2, 4, 7, 6]
    result sequence: [1, 4, 8, 10, 10]

* ``sort_by_key``: performs a key-value sort.
  The algorithm sorts a sequence of keys based on a comparison function object.
  If it is not provided, the elements are compared with ``operator<``.
  A sequence of values is simultaneously permuted according to the sorted order of keys.

>   For correct behavior, a sequence of values must have at least as many elements as the sequence of keys.

Reviewer: I suggest following sentence:
`A sequence of values must have at least as many elements as the sequence of keys, otherwise the behavior is undefined.`

`the behavior is undefined` - "standard" wording on cppreference, f.e., more over we also use one in other parts of the documentation.

Reviewer 2: May be some `assert` required for this check or `exception` ?

Reviewer: We cannot check it, because for "values" we have only "begin"  iterator.

Author: I've applied the suggestion, with slight changes proposed by Alexey below.

    search sequence: [0, 2, 4, 7, 6]
    result sequence: [1, 4, 8, 10, 10]

* ``sort_by_key``: performs a key-value sort. The algorithm sorts a sequence of keys according to

>   a comparison operator. If no comparator is provided, then the elements are compared with ``operator<``.

Reviewer: Probably, should be `std::less` ? 
oneDPL impl uses its own predicate with "operator<", right... But, probably, should we change it as well?

Author: `std::sort` uses `operator<` until c++20, and `std::less` since c++20.
I suppose it does not result in backward compatibility issues since `std::less` is defined as:

> Function object for performing comparisons. The main template invokes operator< on type T.

Namely, `std::less` is a superset including `operator<`.

However, it is a change in the function signature, and it may break the ABI for users. Please, correct me if I am mistaken here.

Author: Ah, just ignore the note about ABI. There is no change in the signature.

Author: There are plenty of algorithms, which should be changed accordingly. Would you mind if it is done separately?

Reviewer: I don't  mind. Lets create an issue to don't forget.

Author: Done: https://github.com/oneapi-src/oneDPL/issues/1779

  A sequence of values is simultaneously permuted according to the sorted order of keys. For correct behavior

>   both sequences shall be of the same size.  

Reviewer: As far as we are changing that sentence... Probably, it makes sense to change `shall be` => `muts be` ? We use `muts be` everywhere in the doc.

Author: I like `must be` more. Done.

    search sequence: [0, 2, 4, 7, 6]
    result sequence: [1, 4, 8, 10, 10]

* ``sort_by_key``: performs a key-value sort. The algorithm sorts a sequence of keys according to
  a comparison operator. If no comparator is provided, then the elements are compared with ``operator<``.
  A sequence of values is simultaneously permuted according to the sorted order of keys. For correct behavior,

>   both sequences must be of the same size.

Reviewer: The given iterator version of `sort_by_key `has 2 iterators for definition of a key sequence and one iterator for definition of a value sequence. 
So, I think the sentence "both sequences must be of the same size" is not quite correct.
For example, we can observe `thrust::sort_by_key ` documentation: 
https://nvidia.github.io/cccl/thrust/api/function_group__sorting_1ga667333ee2e067bb7da3fb1b8ab6d348c.html

Author: That's a good catch. Let me think about the wording. 
The documentation of the additional algorithms does not provide the signature of the functions (which is a huge gap on my opinion), what complicates giving a "crisp" description.

Author: I've corrected the documentation and improved the tests accordingly.

    search sequence: [0, 2, 4, 7, 6]
    result sequence: [1, 4, 8, 10, 10]

* ``sort_by_key``: performs a key-value sort.

>   The algorithm sorts a sequence of keys based on a comparison function object.

Reviewer 3: ```suggestion
  The algorithm sorts a sequence of keys using a given comparison function object.
```

  If it is not provided, the elements are compared with ``operator<``.
  A sequence of values is simultaneously permuted according to the sorted order of keys.

>   For correct behavior, a sequence of values must have at least as many elements as the sequence of keys.

Reviewer 3: ```suggestion
  There must be at least as many values as the number of keys, otherwise the behavior is undefined.
```

    search sequence: [0, 2, 4, 7, 6]
    result sequence: [1, 4, 8, 10, 10]

> * ``stable_sort_by_key``: performs a stable key-value sort. The algorithm sorts the sequence's keys according to 
>   a comparioson operator. If no comparator is provided, then the elements are compared with ``operator<``.
>   The sequence's values are permutated according to the sorted sequence's keys. The prerequisite for correct
>   behavior is that the size for both keys sequence and values sequence shall be the same.  

Reviewer 3: ```suggestion
* ``stable_sort_by_key``: performs a stable key-value sort. The algorithm sorts a sequence of keys according to 
  a comparison operator. If no comparator is provided, then the elements are compared with ``operator<``.
  A sequence of values is simultaneously permuted according to the sorted order of keys. For correct behavior
  both sequences shall be of the same size.  

```

    output_keys:   [0,    0,   3,   3,   4,   5]
    output_values: ['c', 'f', 'a', 'e', 'd', 'b']

  .. note::

     ``sort_by_key`` may implement an unstable sort in the future.
     Use ``stable_sort_by_key`` if the stability is important.

* ``stable_sort_by_key``: performs a stable key-value sort. The algorithm sorts the sequence's keys according to 
  a comparioson operator. If no comparator is provided, then the elements are compared with ``operator<``.
  The sequence's values are permutated according to the sorted sequence's keys. The prerequisite for correct
  behavior is that the size for both keys sequence and values sequence shall be the same.  
  For example::

    keys:   [3,    5,   0,   4,   3,   0]
    values: ['a', 'b', 'c', 'd', 'e', 'f']
    output_keys:   [0,    0,   3,   3,   4,   5]
    output_values: ['c', 'f', 'a', 'e', 'd', 'b']

> 

Reviewer 3: In order to  not duplicate the same information, would not it be better to just change `sort_by_key` to `stable_sort_by_key` in the current wording, and add a short section that describes `sort_by_key` as equivalent to the stable one and adds the caution note?

Author: That's a good idea. It's done.
